### ISSUE TITLE
Newer versions of cloud-init fail to perform schema with DataSourceNotFoundException

### ISSUE THREAD

**Issue by alfred-stokespace** [canonical/cloud-init#4480]
# Bug report
Older versions of cloud-init would allow the `schema` sub command to be run without expecting datasource setup to exist. Newer versions fail with this error...

```
Traceback (most recent call last):
  File "/usr/bin/cloud-init", line 11, in <module>
    load_entry_point('cloud-init==23.2.2', 'console_scripts', 'cloud-init')()
  File "/usr/lib/python3/dist-packages/cloudinit/cmd/main.py", line 1080, in main
    retval = util.log_time(
  File "/usr/lib/python3/dist-packages/cloudinit/util.py", line 2745, in log_time
    ret = func(*args, **kwargs)
  File "/usr/lib/python3/dist-packages/cloudinit/config/schema.py", line 1267, in handle_schema_args
    paths = read_cfg_paths(fetch_existing_datasource="trust")
  File "/usr/lib/python3/dist-packages/cloudinit/cmd/devel/__init__.py", line 30, in read_cfg_paths
    init.fetch(existing=fetch_existing_datasource)
  File "/usr/lib/python3/dist-packages/cloudinit/stages.py", line 436, in fetch
    return self._get_data_source(existing=existing)
  File "/usr/lib/python3/dist-packages/cloudinit/stages.py", line 327, in _get_data_source
    (ds, dsname) = sources.find_source(
  File "/usr/lib/python3/dist-packages/cloudinit/sources/__init__.py", line 1023, in find_source
    raise DataSourceNotFoundException(msg)
cloudinit.sources.DataSourceNotFoundException: Did not find any data source, searched classes: (DataSourceNone)
```

## Steps to reproduce the problem
Ensure you're using a recent cloud init like `cloud-init==23.2.2` and attempt a simple schema validation like so...

config file `test.yaml`
containing 
```
#cloud-config
autoinstall:
  version: 1
  identity:
    hostname: ubuntu-server
    password: "$6$exDY1mhS4KUYCE/2$zmn9ToZwTKLhCw.b4/b.ZRTIZM30JZ4QrOQ2aOXJ8yk96xpcCof0kxKwuX1kqLG/ygbJ1f8wxED22bTL4F46P0"
    username: ubuntu
  interactive-sections:
    - storage
```
and issue command 
` cloud-init schema --config-file test.yaml`
You will get the error above

NOW downgrade ...
```
 sudo apt-get install cloud-init=23.1.2-0ubuntu0~20.04.2
```
issue the same command on the same file

` cloud-init schema --config-file test.yaml`
and you get the good stuff!
```
Valid cloud-config: test.yaml
```
## Environment details
- Cloud-init version: `23.2.2`
- Operating System Distribution: Ubuntu 20.04
- Cloud provider, platform or installer type: N/A

## cloud-init logs
See above

**Comment by aciba90** [comment]
Thanks for reporting this issue.

This problem was introduced in [23.2](https://github.com/canonical/cloud-init/releases/tag/23.2) by https://github.com/canonical/cloud-init/commit/73f34575da4f99a376998516a13c3a79cc640ae3 and it has been fixed in https://github.com/canonical/cloud-init/commit/893b89a974d473b02dcdcc47074c0f04e472abf9 which is not included in the latest upstream release [23.3.1](https://github.com/canonical/cloud-init/releases/tag/23.3.1).

**Comment by aciba90** [comment]
As a workaround to avoid this issue, one can provide [datasource configuration](https://canonical-cloud-init.readthedocs-hosted.com/en/latest/explanation/kernel-cmdline.html#datasource-discovery-override) to fallback to `DataSourceNone` and optionally re-run [datasource detection](https://canonical-cloud-init.readthedocs-hosted.com/en/latest/howto/identify_datasource.html#how-can-i-re-run-datasource-detection-and-cloud-init).

As said previously, the issue is believed to be fixed by https://github.com/canonical/cloud-init/commit/893b89a974d473b02dcdcc47074c0f04e472abf9, so I am going to mark this as closed.

Please, re-open / comment on this ticket if that is not the case, thanks.

**Comment by blackboxsw** [comment]
Expectation is this fix will be part of 23.4 release (scheduled mid-November) and in the meantime.

Another alternative for obtaining this fix if this is a development environment with cloud-init disabled:
  One could install latest cloud-init deb package via `add-apt-repository ppa:cloud-init-dev/daily; apt install cloud-init`. 

Note that this version of cloud-init is a daily copy of tip of main with no validation, so if you want cloud-init active in this environment across reboots it's probbab
